### PR TITLE
make HMODULE equal to the same opaque resource as HINSTANCE

### PR DESCRIPTION
**Pull request by expikr**
According to [MSDN](https://learn.microsoft.com/en-us/windows/win32/winprog/windows-data-types), they differ only in 16bit windows, but in fact point to the same opaque resources in non-16bit versions. 

I’ve also taken the liberty to change UINT and INT from c_uint and c_int to u32 and i32, since MSDN states explicitly that they are 32 bit, as well as lowering their dependents to their generic counterparts. 

Other changes made according to MSDN:

* LANGID changed to `u16` ([“Each language has a unique language identifier (data type LANGID), a 16-bit value that consists of a primary language identifier and a sublanguage identifier.”](https://learn.microsoft.com/en-us/windows/win32/intl/language-identifiers))
* ~~HLOCAL changed to `*anyopaque` (“A handle to a local memory block” which can be any data type in the destination, ergo a type-erased pointer)~~ [see comment](https://github.com/ziglang/zig/pull/20231#discussion_r1632948034)

### CODE REVIEW

**Review by marler8997 — APPROVED** [review]
This PR brings std closer in line with zigwin32 (and win32metadata).

The following types actually don't appear in zigwin32/win32metadata at all, the definitions are used directly instead:

```
HMODULE (HINSTANCE used instead)
INT (i32 used instead)
UINT (u32 used instead)
DWORD_PTR (usize used instead)
HLOCAL (isize used instead, isize looks wrong though, using *anyopaque like this PR does makes more sense I think)
LANGID (no equivalent, this is part of a macro and win32metadata doesn't include macros)
```

I'd probably just remove them but that would require alot more changes.  The PR as is is a step in the right direction either way.

The rest of the types modified in this PR now match how zigwin32 defines them.

File: lib/std/os/windows.zig [hunk 1]
```diff
@@ -2859,12 +2859,12 @@ pub const LONG = i32;
 pub const ULONG64 = u64;
 pub const ULONGLONG = u64;
 pub const LONGLONG = i64;
-pub const HLOCAL = HANDLE;
-pub const LANGID = c_ushort;
+pub const HLOCAL = *anyopaque;
```
Comment (andrewrk) [inline lib/std/os/windows.zig:2862]: this change in not justified in your commit descriptions

**Review by andrewrk — CHANGES_REQUESTED** [review]
please make only one commit and make the commit description justify your changes

File: lib/std/os/windows.zig (hunk 1 above)
Comment (marler8997) [reply lib/std/os/windows.zig:2862]: I think `*opaque{}` makes the most sense here and matches what win32metadata is currently doing (zigwin32 is out-of-date and this has changed since).  win32metadta has defined its own Handle type for `LocalAlloc`.  The result of `LocalAlloc` isn't meant to be used anywhere `HANDLE` is meant to be used as far as I know.

That being said, another alternative is to remove `HLOCAL` and `LocalFree`, doesn't seem like zig is using them.

P.S. I see you updated your PR description but I believe Andrew was wanting you to put your comments in the commit description (people's clones of the git don't have access to github PR descriptions).

Comment (daurnimator) [reply lib/std/os/windows.zig:2862]: @expikr FWIW, to be "more-proper", we should have a way to access/twiddle the lower 2 bits.
In a HANDLE and other such kernel handles, the lowest 2 bits are ignored and can be used by the user/program to "tag" various handles going in or out of the kernel.
So a more correct type might be:
```zig
packed struct {
  bits: u2,
  opaque_bits: std.meta.Int(.signed, @bitSizeOf(usize)-2),
}
```
or similar?

Comment (daurnimator) [reply lib/std/os/windows.zig:2862]: > but Windows have other objects that have `H****` handle types:
> 
> > Not very well known is that the bottom two bits of kernel HANDLEs are always zero; in other words, their numeric value is always a multiple of 4. Note that this applies only to kernel HANDLEs; it does not apply to pseudo-handles or to any other type of handle (USER handles, GDI handles, multimedia handles…) Kernel handles are things you can pass to[ the CloseHandle function](http://msdn.microsoft.com/library/en-us/sysinfo/base/closehandle.asp).
> > — https://devblogs.microsoft.com/oldnewthing/20050121-00/?p=36633

Yes indeed e.g. GDI handles exist.
Perhaps we should have something like:
```zig
pub const HANDLE = extern union {
    pub const KERNEL = packed struct {
      bits: u2,
      opaque_bits: std.meta.Int(.signed, @bitSizeOf(usize)-2),
    };
    kernel: KERNEL,
    user: *anyopaque,
    gdi: *anyopaque,
}
```

And then e.g. the `CloseHandle ` function mentioned would take a `HANDLE.KERNEL` instead of just a HANDLE?

Comment (squeek502) [reply lib/std/os/windows.zig:2862]: I don't have enough practical experience with `HLOCAL` to have a worthwhile opinion here.

One data point that might be relevant, though, is how these handles are defined with [`STRICT`](https://learn.microsoft.com/en-us/windows/win32/winprog/strict-type-checking) enabled. From my reading of `windef.h`, `HLOCAL` bypasses `STRICT` and is always the same type as `HANDLE` (i.e. it is defined as `typedef HANDLE HLOCAL;` instead of using the `DECLARE_HANDLE` macro which gives a handle a distinct type when `STRICT` is enabled).

Comment (squeek502) [reply lib/std/os/windows.zig]: > makein HCURSOR a synonym of HICON

This is one I think I'd actually recommend against unless there's a particular use case in mind. You're right that `WinDef.h` defines them as aliases, and they may be able to be used interchangeably in certain circumstances, but given that there are things like separate [DestroyIcon](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-destroyicon) and [DestroyCursor](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-destroycursor) functions, enforcing type safety seems like a good idea.

File: lib/std/os/windows.zig [hunk 2]
```diff
@@ -2859,12 +2849,27 @@ pub const LONG = i32;
...
 pub const LONGLONG = i64;
-pub const HLOCAL = HANDLE;
-pub const LANGID = c_ushort;
+pub const LANGID = u16;
+
+// strict handles and their synonyms
+pub const HINSTANCE = *opaque {};
+pub const HMODULE = HINSTANCE;
+pub const HICON = *opaque {};
+pub const HCURSOR = *opaque {}; // actually equal to HICON
```
Comment (andrewrk) [inline lib/std/os/windows.zig:2858]: if it's actually equal to HICON then it should look like this:

```suggestion
pub const HCURSOR = HICON;
```

**Review by andrewrk — CHANGES_REQUESTED** [review]

File: lib/std/os/windows.zig (hunk 2 above)
Comment (squeek502) [reply lib/std/os/windows.zig:2858]: I think the comment should be removed. I suggested not making them aliases here: https://github.com/ziglang/zig/pull/20231#discussion_r1639539907

File: lib/std/os/windows.zig [hunk 3]
```diff
@@ -2859,12 +2849,27 @@ pub const LONG = i32;
...
 pub const LONGLONG = i64;
-pub const HLOCAL = HANDLE;
-pub const LANGID = c_ushort;
+pub const LANGID = u16;
+
+// strict handles and their synonyms
+pub const HINSTANCE = *opaque {};
+pub const HMODULE = HINSTANCE;
+pub const HICON = *opaque {};
+pub const HCURSOR = *opaque {}; // distinct from HICON to prefer destructor type safety, but windef.h actually makes them equal even under STRICT mode
```
Comment (andrewrk) [inline lib/std/os/windows.zig:2858]: why is it not a doc comment

**Review by andrewrk — CHANGES_REQUESTED** [review]

**Comment by mochalins** [comment]
Would love these changes merged in; do you have any more updates planned to this PR? @expikr  
If I understand Andrew's [comment](https://github.com/ziglang/zig/pull/20231#pullrequestreview-2105923199) correctly, I believe having the 4 commits currently in this PR squashed into one commit with a detailed commit description would be preferred (and hopefully get this landed faster?).

File: lib/std/os/windows.zig [hunk 4]
```diff
@@ -2859,12 +2839,66 @@ pub const LONG = i32;
 pub const ULONG64 = u64;
 pub const ULONGLONG = u64;
 pub const LONGLONG = i64;
-pub const HLOCAL = HANDLE;
-pub const LANGID = c_ushort;
+pub const LANGID = u16;
+
+/// pointer to immutable null-terminated u8 string
```
Comment (andrewrk) [inline lib/std/os/windows.zig:2844]: > pointer

redundant

> immutable

redundant

> null-terminated

redundant

> u8

this is not an appropriate place to mention what `CHAR` is aliased to

> string

meaningless at best, misleading at worst

File: lib/std/os/windows.zig [hunk 5]
```diff
@@ -2005,10 +2005,6 @@ pub fn HeapDestroy(hHeap: HANDLE) void {
     assert(kernel32.HeapDestroy(hHeap) != 0);
 }
 
-pub fn LocalFree(hMem: HLOCAL) void {
-    assert(kernel32.LocalFree(hMem) == null);
-}
-
```
Comment (andrewrk) [inline lib/std/os/windows.zig:2011]: this breaking change should be a separate PR. Your PRs require a lot of reviews, so please limit yourself to 1 open PR at a time. you're already 5 over this quota.

**Review by andrewrk — CHANGES_REQUESTED** [review]

**Review by andrewrk — CHANGES_REQUESTED** [review]
Some excerpts from `zig translate-c` of windows.h using -target x86_64-windows-gnu:

```
pub const BOOLEAN = BYTE;
// ...
pub const HCRYPTPROV = ULONG_PTR;
// ...
pub const INT = c_int;
```

I don't see any reasoning given for these changes.

File: lib/std/os/windows.zig [hunk 6]
```diff
@@ -2809,43 +2809,27 @@ pub const WINAPI: std.builtin.CallingConvention = if (native_arch == .x86)
 else
     .C;
 
+// primitive types
```
Comment (andrewrk) [inline lib/std/os/windows.zig:2812]: not helpful

File: lib/std/os/windows.zig [hunk 7]
```diff
@@ -2859,12 +2843,55 @@ pub const LONG = i32;
...
+pub const HDC = *opaque {};
+
+/// Handle to OpenGL rendering context.
+pub const HGLRC = *opaque {};
+
+// polymorphic handles
+pub const HANDLE = *anyopaque;
+pub const HLOCAL = *anyopaque;
 
+// polymorphic types
```
Comment (andrewrk) [inline lib/std/os/windows.zig:2891]: not helpful

File: lib/std/os/windows.zig [hunk 8]
```diff
@@ -2859,12 +2843,55 @@ pub const LONG = i32;
...
+/// Handle to window object.
+pub const HWND = *opaque {};
+
+/// Handle to device context.
+pub const HDC = *opaque {};
+
+/// Handle to OpenGL rendering context.
+pub const HGLRC = *opaque {};
+
+// polymorphic handles
```
Comment (andrewrk) [inline lib/std/os/windows.zig:2887]: not helpful

**Comment by andrewrk** [comment]
In https://github.com/ziglang/zig/pull/20231#discussion_r1649777759 I asked you to limit yourself to 1 open pull request at a time. You have not done as I asked. To show you that I am serious I will now close all your open PRs and block you from the org for 1 year. You are welcome to try again after that.
